### PR TITLE
feat: add elasticsearch steps

### PR DESCRIPTION
**Pull request by ismtabo**

### CODE REVIEW

**Review by cbrea — APPROVED** [review]
LGTM

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -22,7 +22,18 @@ jobs:
...
+      elasticsearch:
+        image: elasticsearch:7.12.0
+        ports:
+          - 9200:9200
+        options: >-
+          -e="discovery.type=single-node"
+          --health-cmd "curl http://localhost:9200/_cluster/health"
+          --health-interval 10s
+          --health-timeout 5s
+          --health-retries 10
```
Comment (mariolg) [inline .github/workflows/ci.yml:35]: redis does only 5 retries, make it homogeneous.

File: steps/elasticsearch/session.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+// Session contains the information of a elasticsearch session.
+type Session struct {
+	Client       *elasticsearch.Client
+	SearchResult golium.Map
+	Correlator   string
+}
+
+// ConfigureClient creates a elasticsearch connection based on the URI.
+func (s *Session) ConfigureConnection(ctx context.Context, config elasticsearch.Config) error {
```
Comment (mariolg) [inline steps/elasticsearch/session.go:40]: I guess here you're configuring a client (NewClient invoked).

File: steps/elasticsearch/session.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	var err error
+	if s.Client, err = elasticsearch.NewClient(config); err != nil {
+		return errors.Wrap(err, "failed configuring elasticsearch client")
+	}
+	s.Correlator = uuid.NewString()
+	return nil
+}
+
+// CreatesDocument creates a document in index with given JSON properties.
```
Comment (mariolg) [inline steps/elasticsearch/session.go:49]: CreateDocument  creates a document, with the given JSON and for the given index.

Comment (mariolg) [reply steps/elasticsearch/session.go:49]: Is not better `NewDocument`?

File: steps/elasticsearch/session.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+func (s *Session) CreatesDocument(ctx context.Context, index string, props map[string]interface{}) error {
+	logger := GetLogger()
+	var err error
+	data := ""
+	for key, value := range props {
+		data, err = sjson.Set(data, key, value)
```
Comment (mariolg) [inline steps/elasticsearch/session.go:55]: I think you don't need external library `sjson`, have a look at: https://play.golang.org/p/goJcM7FXWFp

File: steps/elasticsearch/steps.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/Telefonica/golium"
+	"github.com/cucumber/godog"
+	"github.com/elastic/go-elasticsearch/v8"
+)
+
+// Steps type is responsible to initialize the HTTP client steps in godog framework.
```
Comment (mariolg) [inline steps/elasticsearch/steps.go:26]: HTTP?

File: steps/elasticsearch/steps.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+type Steps struct {
+}
+
+// InitializeSteps adds client HTTP steps to the scenario context.
+// It implements StepsInitializer interface.
+// It returns a new context (context is immutable) with the HTTP Context.
+func (s Steps) InitializeSteps(ctx context.Context, scenCtx *godog.ScenarioContext) context.Context {
+	// Initialize the HTTP session in the context
```
Comment (mariolg) [inline steps/elasticsearch/steps.go:34]: HTTP?

**Review by mariolg — CHANGES_REQUESTED** [review]
Have a look at my comments.

File: steps/elasticsearch/session.go (hunk 4 above)
Comment (ismtabo) [reply steps/elasticsearch/session.go:55]: With `sjson` we keep the same behaviour such as in the other steps, keeping some addons as the [path syntax][1].


[1]: https://github.com/tidwall/sjson#path-syntax

**Review by mariolg — APPROVED** [review]
LGTM

**Review by cbrea — APPROVED** [review]
LGTM

File: steps/elasticsearch/session.go [hunk 7]
```diff
@@ -69,6 +71,11 @@ func (s *Session) NewDocument(ctx context.Context, index string, props map[strin
 	if res.IsError() {
 		return s.parseErrorResponse(res)
 	}
+	if document, err := s.getResponseIndexedDocument(res); err == nil {
+		s.indexedDocuments = append(s.indexedDocuments, document)
+	} else {
+		logger.LogError(errors.Wrap(err, "failed storeing indexed document"), s.Correlator)
```
Comment (mariolg) [inline steps/elasticsearch/session.go:77]: typo: `storing`

File: steps/elasticsearch/session.go [hunk 8]
```diff
@@ -107,6 +114,48 @@ func (s *Session) ValidateDocumentJSONProperties(ctx context.Context, props map[
...
+			esutil.BulkIndexerItem{
+				Action:     "delete",
+				Index:      document.Index,
+				DocumentID: document.Id,
+				OnFailure: func(ctx context.Context, item esutil.BulkIndexerItem, res esutil.BulkIndexerResponseItem, err error) {
+					errMsg := fmt.Sprintf("failed deleting document to clean up indexes for document %+v", document)
+					if err != nil {
+						logger.LogError(errors.Wrap(err, errMsg), s.Correlator)
+					} else {
+						logger.LogError(errors.Wrap(errors.Errorf("elasticsearch error: %s: %s", res.Error.Type, res.Error.Reason), errMsg), s.Correlator)
```
Comment (mariolg) [inline steps/elasticsearch/session.go:139]: Better `elasticsearch error. %s: %s`

**Review by mariolg — COMMENTED** [review]
LGTM

**Review by cbrea — APPROVED** [review]
LGTM

**Pull request merged**